scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+        """
+
+        num_cpus = cpu_count()
+        # we need one main thread + N worker threads, so `N = num_cpus - 1`,
+        # and we need at least 1 worker thread.
+        num_workers = num_cpus - 1 if num_cpus > 0 else 1
+        with ThreadPoolExecutor(max_workers=num_workers) as p:
+            task_queue: Queue[_T] = Queue()
+            done_queue: Queue[_T] = Queue()
+            shutdown: EventClass = Event()
+
+            def worker(n: int) -> None:
+                # Attempt to grab a task from the queue,
+                # execute it, then put it in the done queue.
+                while not shutdown.is_set():
+                    try:
+                        node = task_queue.get_nowait()
+                        f(node)
+                        done_queue.put(node)
+                    except Empty:
+                        time.sleep(0)  # yield to prevent busy-looping
+                        continue
+
+            for n in range(0, num_workers):  # start all workers
+                p.submit(worker, n)
+
+            tokens = max_tokens
+            in_progress = 0
+            last_refill = time.time()
+            while not self._is_done():
+                # This loop has two parts, `push` and `pull`.
+                #
+                # The `push` loop attempts to push tasks
+                # onto the `task_queue` while there are
+                # some tasks ready to go, and some tokens left.
+                # It is also responsible for refreshing the bucket.
+                #
+                # The `pull` loop attempts to retrieve done
+                # tasks and decrement the dependency counter on
+                # their dependents.
+                #
+                # Once a node has no pending dependencies left,
+                # it becomes ready and will be queued in one of
+                # the iterations of the `push` loop.
+                #
+                # It's important to always use non-blocking `get`
+                # with the task and done queues.
+                # This is so that both the push and pull loops
+                # can eventually make progress.
+                while len(self._queue) > 0:  # push loop
+                    now = time.time()
+                    if now - last_refill > refill_interval_s:
+                        tokens = max_tokens - in_progress
+                        last_refill = now
+
+                    if len(self._queue) == 0 or tokens == 0:
+                        break
+
+                    tokens -= 1
+                    in_progress += 1
+                    task_queue.put(self._queue.pop())
+
+                try:
+                    while True:  # pull loop
+                        self._finish(done_queue.get_nowait())
+                        in_progress -= 1
+                except Empty:
+                    time.sleep(0)  # yield here to prevent busy-looping
+
+            shutdown.set()
```

Comment:
we could add a sanity-check here in the end, e.g. that `counter` is zero for every job, and that the number of executed jobs is the same as the number of nodes in the graph, etc.

btw, is it even possible to call `walk` twice in a row? Seems like `counter` is set up in the constructor when it probably should be set up in the start of `walk` instead

Reply:
Added the sanity check, and made it possible to call `walk_parallel` multiple times.